include/kernel.h

```diff
@@ -4430,6 +4430,56 @@ __syscall int k_pipe_get(struct k_pipe *pipe, void *data,
 extern void k_pipe_block_put(struct k_pipe *pipe, struct k_mem_block *block,
 			     size_t size, struct k_sem *sem);
 
+/**
+ * @brief Query the number of bytes that may be read from @a pipe.
+ *
+ * @param pipe Address of the pipe.
+ *
+ * @retval a number n such that 0 <= n <= @ref k_pipe.size; the
+ *         result is zero for unbuffered pipes.
+ */
+__syscall size_t k_pipe_read_avail(const struct k_pipe *pipe);
+
+static inline size_t z_impl_k_pipe_read_avail(const struct k_pipe *pipe)
+{
+	if (pipe->buffer == NULL || pipe->size == 0) {
+		return 0;
+	}
+
+	if (pipe->read_index == pipe->write_index) {
+		return pipe->bytes_used;
+	} else if (pipe->read_index < pipe->write_index) {
+		return pipe->write_index - pipe->read_index;
```

Comment:
Isn't the pipe a ring?  I think this subtraction needs to be modulus the size, no?

Reply:
That's the "invisible else" condition, but we avoid the modulus by changing the order of the operands.

Reply:
It is a ring. If read < write, then it's the straight difference. Otherwise, we would use the modulus operator, but we can get away from a potentially expensive division in hw by subtracting from total size.
The read == write is a special case.
I wish there was a way to do this without branching (i.e. stalling the pipeline). Bit-twiddling hacks can be pretty awesome.